cpp/skyweaver/detail/IncoherentDedispersionPipeline.cu

```diff
@@ -79,7 +80,44 @@ void IncoherentDedispersionPipeline<InputType, OutputType, Handler>::
     BOOST_LOG_TRIVIAL(debug) << "Passing output buffer to handler: "
                              << _output_buffers[ref_dm_idx].describe();
     _timer.start("file writing");
-    _handler(_output_buffers[ref_dm_idx], ref_dm_idx);
+
+    if (_n_tdb_files > 1)
+    {
+        const std::size_t ndms     = _output_buffers[ref_dm_idx].ndms();
+        const std::size_t nbeams   = _output_buffers[ref_dm_idx].nbeams();
+        const std::size_t nsamples = _output_buffers[ref_dm_idx].nsamples();
+        const std::size_t nbeams_per_file = _config.nbeams_per_file();
+
+        _beamsplit_buffer.metalike(_output_buffers[ref_dm_idx]);
+        _beamsplit_buffer.resize({nsamples, ndms, nbeams_per_file});
+
+        std::size_t td_input_offset;
+        std::size_t td_output_offset;
+        std::size_t b_offset;
+
+        for (std::size_t tdb_file_idx = 0; tdb_file_idx < _n_tdb_files; ++tdb_file_idx)
+        {
+            b_offset = tdb_file_idx * nbeams_per_file;
+
+            for (std::size_t tdidx = 0; tdidx < nsamples * ndms; ++tdidx)
+            {
+                td_input_offset = tdidx * nbeams;
+                td_output_offset = tdidx * nbeams_per_file;
+
+                for (std::size_t bidx = 0; bidx < nbeams_per_file ; ++bidx)
```

Comment:
Probably more efficient to swap this out with a std::copy for the inner set of beams. Might be easier to get the compiler to optimise this.